### PR TITLE
Search any occurrence of substring in STRPOS function

### PR DESCRIPTION
**Pull request by guyco33**
Add `instance` parameter to `strpos` function to return the position of the nth occurrence of substring in a string.
When `instance` is negative number, the string is searched from the end.

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java [hunk 1]
```diff
@@ -195,14 +196,70 @@ public static Slice reverse(@SqlType("varchar(x)") Slice slice)
...
+        }
         if (substring.length() == 0) {
             return 1;
         }
 
-        int index = string.indexOf(substring);
-        if (index < 0) {
-            return 0;
+        int foundInstances = 0;
+        // set the initial index just before the start of the string
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java]: These comments are not very useful, as they just describe what the code is doing. As a general rule, document what's not obvious from the immediate code (hidden relationships, assumptions, non-obvious behaviors, etc).

File: presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java [hunk 2]
```diff
@@ -195,14 +196,70 @@ public static Slice reverse(@SqlType("varchar(x)") Slice slice)
...
+    {
+        if (instance <= 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "'instance' must be a positive or negative number.");
+        }
+        if (substring.length() == 0) {
+            return 1;
+        }
+
+        int foundInstances = 0;
+        // set the initial index just after the end of the string
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java]: Same here

File: presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java [hunk 3]
```diff
@@ -195,14 +196,70 @@ public static Slice reverse(@SqlType("varchar(x)") Slice slice)
...
+        // this is to allow for the initial index decrement
+        int index = string.length();
+        do {
+            // step backwards through string
+            index = string.toStringUtf8().lastIndexOf(substring.toStringUtf8(), index - 1);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java]: This will get the last index in terms of String characters. But we use this index in `countCodePoints(string, 0, index) + 1`, which is based on bytes.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestStringFunctions.java [hunk 4]
```diff
@@ -290,6 +290,45 @@ public void testStringPosition()
...
+        assertFunction("STRPOS(NULL, '')", BIGINT, null);
+        assertFunction("STRPOS('', NULL)", BIGINT, null);
+        assertFunction("STRPOS(NULL, NULL)", BIGINT, null);
+        assertInvalidFunction("STRPOS('abc/xyz/foo/bar', '/', 0)", "'instance' must be a positive or negative number.");
+        assertInvalidFunction("STRPOS('', '', 0)", "'instance' must be a positive or negative number.");
+
+        assertFunction("STRPOS('abc/xyz/foo/bar', '/')", BIGINT, 4L);
+        assertFunction("STRPOS('\u4FE1\u5FF5,\u7231,\u5E0C\u671B', '\u7231')", BIGINT, 4L);
+        assertFunction("STRPOS('\u4FE1\u5FF5,\u7231,\u5E0C\u671B', '\u5E0C\u671B')", BIGINT, 6L);
+        assertFunction("STRPOS('\u4FE1\u5FF5,\u7231,\u5E0C\u671B', 'nice')", BIGINT, 0L);
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestStringFunctions.java]: Can you add tests for searching from the back for strings containing non-ASCII unicode characters?

File: presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java (hunk 3 above)
Comment (guyco33) [reply presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java]: Thanks! I didn't catch it since I skipped the tests with non-ASCII unicode characters

**Comment by martint** [comment]
Merged, thanks for your contribution!
